File: ClassyTaxiJava/app/src/test/java/com/sample/android/classytaxijava/billing/BillingUtilitiesTest.java
package com.sample.android.classytaxijava.billing;


import static com.sample.android.classytaxijava.billing.BillingUtilities.getSubscriptionForSku;
import static com.sample.android.classytaxijava.billing.BillingUtilities.isAccountHold;
import static com.sample.android.classytaxijava.billing.BillingUtilities.isBasicContent;
import static com.sample.android.classytaxijava.billing.BillingUtilities.isGracePeriod;
import static com.sample.android.classytaxijava.billing.BillingUtilities.isPaused;
import static com.sample.android.classytaxijava.billing.BillingUtilities.isPremiumContent;
import static com.sample.android.classytaxijava.billing.BillingUtilities.isSubscriptionRestore;
import static com.sample.android.classytaxijava.billing.BillingUtilities.isTransferRequired;

import com.sample.android.classytaxijava.data.SubscriptionStatus;

import org.junit.Test;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.nullValue;

import java.util.ArrayList;
import java.util.List;

public class BillingUtilitiesTest {

    @Test
    public void subscriptionForSku_returnsSubscriptionForExistingSku() {
        SubscriptionStatus basicSub = new SubscriptionStatus();
        basicSub.setSku("basic_subscription");
        SubscriptionStatus premiumSub = new SubscriptionStatus();
        premiumSub.setSku("premium_subscription");

        List<SubscriptionStatus> noSkuSubscriptionList = new ArrayList<SubscriptionStatus>();
        noSkuSubscriptionList.add(basicSub);
        noSkuSubscriptionList.add(premiumSub);

        String sku = "basic_subscription";
        SubscriptionStatus result = getSubscriptionForSku(noSkuSubscriptionList, sku);
        assertThat(result.getSku(), is(sku));
    }

    @Test
    public void subscriptionForSku_returnsNullForNonExistingSku() {
        SubscriptionStatus basicSub = new SubscriptionStatus();
        basicSub.setSku("basic_subscription");
        SubscriptionStatus premiumSub = new SubscriptionStatus();
        premiumSub.setSku("premium_subscription");

        List<SubscriptionStatus> noSkuSubscriptionList = new ArrayList<SubscriptionStatus>();
        noSkuSubscriptionList.add(basicSub);
        noSkuSubscriptionList.add(premiumSub);

        String sku = "non_cheap_subscription";
        SubscriptionStatus result = getSubscriptionForSku(noSkuSubscriptionList, sku);
        assertThat(result, is(nullValue()));

    }

    @Test
    public void isGracePeriod_returnsTrueIfInGracePeriod() {
        SubscriptionStatus onGraceSubscription = new SubscriptionStatus();
        onGraceSubscription.setSubAlreadyOwned(false);
        onGraceSubscription.setEntitlementActive(true);
        onGraceSubscription.setGracePeriod(true);
        Boolean result = isGracePeriod(onGraceSubscription);
        assertThat(result, is(true));
    }

    @Test
    public void isGracePeriod_returnsFalseIfNotInGracePeriod() {
        SubscriptionStatus notOnGraceSubscription = new SubscriptionStatus();
        notOnGraceSubscription.setSubAlreadyOwned(false);
        notOnGraceSubscription.setEntitlementActive(true);
        notOnGraceSubscription.setGracePeriod(false);
        Boolean result = isGracePeriod(notOnGraceSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isGracePeriod_returnsFalseIfAlreadyOwnedSub() {
        SubscriptionStatus alreadyOwnedSubscription = new SubscriptionStatus();
        alreadyOwnedSubscription.setSubAlreadyOwned(true);
        alreadyOwnedSubscription.setEntitlementActive(true);
        alreadyOwnedSubscription.setGracePeriod(true);
        Boolean result = isGracePeriod(alreadyOwnedSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isGracePeriod_returnsFalseIfEntitlementInactive() {
        SubscriptionStatus entitlementInactiveSubscription = new SubscriptionStatus();
        entitlementInactiveSubscription.setSubAlreadyOwned(false);
        entitlementInactiveSubscription.setEntitlementActive(false);
        entitlementInactiveSubscription.setGracePeriod(true);
        Boolean result = isGracePeriod(entitlementInactiveSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isGracePeriod_returnsFalseForNullSub() {
        SubscriptionStatus nullSubscription = null;
        Boolean result = isGracePeriod(nullSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isSubscriptionRestore_returnsTrueForNonRenewingSub() {
        SubscriptionStatus willNotRenewSubscription = new SubscriptionStatus();
        willNotRenewSubscription.setSubAlreadyOwned(false);
        willNotRenewSubscription.setEntitlementActive(true);
        willNotRenewSubscription.setWillRenew(false);
        Boolean result = isSubscriptionRestore(willNotRenewSubscription);
        assertThat(result, is(true));

    }

    @Test
    public void isSubscriptionRestore_returnsFalseForRenewingSub() {
        SubscriptionStatus willRenewSubscription = new SubscriptionStatus();
        willRenewSubscription.setSubAlreadyOwned(false);
        willRenewSubscription.setEntitlementActive(true);
        willRenewSubscription.setWillRenew(true);
        Boolean result = isSubscriptionRestore(willRenewSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isSubscriptionRestore_returnsFalseForNullSub() {
        SubscriptionStatus nullSubscription = null;
        Boolean result = isSubscriptionRestore(nullSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isBasicContent_returnsTrueForBasicSku() {
        SubscriptionStatus basicSubscription = new SubscriptionStatus();
        basicSubscription.setSubAlreadyOwned(false);
        basicSubscription.setEntitlementActive(true);
        basicSubscription.setSku("basic_subscription");
        Boolean result = isBasicContent(basicSubscription);
        assertThat(result, is(true));
    }

    @Test
    public void isBasicContent_returnsFalseForNonBasicSku() {
        SubscriptionStatus nonBasicSubscription = new SubscriptionStatus();
        nonBasicSubscription.setSubAlreadyOwned(false);
        nonBasicSubscription.setEntitlementActive(true);
        nonBasicSubscription.setSku("premium_subscription");
        Boolean result = isBasicContent(nonBasicSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isBasicContent_returnsFalseForNullSub() {
        SubscriptionStatus nullSubscription = null;
        Boolean result = isBasicContent(nullSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isPremiumContent_returnsTrueForPremiumSku() {
        SubscriptionStatus premiumSubscription = new SubscriptionStatus();
        premiumSubscription.setSubAlreadyOwned(false);
        premiumSubscription.setEntitlementActive(true);
        premiumSubscription.setSku("premium_subscription");
        Boolean result = isPremiumContent(premiumSubscription);
        assertThat(result, is(true));
    }

    @Test
    public void isPremiumContent_returnsFalseForNonPremiumSku() {
        SubscriptionStatus nonPremiumSubscription = new SubscriptionStatus();
        nonPremiumSubscription.setSubAlreadyOwned(false);
        nonPremiumSubscription.setEntitlementActive(true);
        nonPremiumSubscription.setSku("basic_subscription");
        Boolean result = isPremiumContent(nonPremiumSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isPremiumContent_returnsFalseForNullSub() {
        SubscriptionStatus nullSubscription = null;
        Boolean result = isPremiumContent(nullSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isAccountHold_returnsTrueForOnHoldSub() {
        SubscriptionStatus onHoldSubscription = new SubscriptionStatus();
        onHoldSubscription.setSubAlreadyOwned(false);
        onHoldSubscription.setEntitlementActive(false);
        onHoldSubscription.setAccountHold(true);
        Boolean result = isAccountHold(onHoldSubscription);
        assertThat(result, is(true));
    }

    @Test
    public void isAccountHold_returnsFalseForNonHoldSub() {
        SubscriptionStatus notHoldSubscription = new SubscriptionStatus();
        notHoldSubscription.setSubAlreadyOwned(false);
        notHoldSubscription.setEntitlementActive(true);
        notHoldSubscription.setAccountHold(false);
        Boolean result = isAccountHold(notHoldSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isAccountHold_returnsFalseForNullSub() {
        SubscriptionStatus nullSubscription = null;
        Boolean result = isAccountHold(nullSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isPaused_returnsTrueForPausedSub() {
        SubscriptionStatus pausedSubscription = new SubscriptionStatus();
        pausedSubscription.setSubAlreadyOwned(false);
        pausedSubscription.setEntitlementActive(false);
        pausedSubscription.setPaused(true);
        Boolean result = isPaused(pausedSubscription);
        assertThat(result, is(true));
    }

    @Test
    public void isPaused_returnsFalseForNonPausedSub() {
        SubscriptionStatus nonPausedSubscription = new SubscriptionStatus();
        nonPausedSubscription.setSubAlreadyOwned(false);
        nonPausedSubscription.setEntitlementActive(true);
        nonPausedSubscription.setPaused(false);
        Boolean result = isPaused(nonPausedSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isPaused_returnsFalseForNullSub() {
        SubscriptionStatus nullSubscription = null;
        Boolean result = isPaused(nullSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isTransferRequired_returnsTrueForAlreadyOwnedSub() {
        SubscriptionStatus alreadyOwnedSubscription = new SubscriptionStatus();
        alreadyOwnedSubscription.setSubAlreadyOwned(true);
        Boolean result = isTransferRequired(alreadyOwnedSubscription);
        assertThat(result, is(true));
    }

    @Test
    public void isTransferRequired_returnsFalseForNotAlreadyOwnedSub() {
        SubscriptionStatus notAlreadyOwnedSubscription = new SubscriptionStatus();
        notAlreadyOwnedSubscription.setSubAlreadyOwned(false);
        Boolean result = isTransferRequired(notAlreadyOwnedSubscription);
        assertThat(result, is(false));
    }

    @Test
    public void isTransferRequired_returnsFalseForNullSub() {
        SubscriptionStatus nullSubscription = null;
        Boolean result = isTransferRequired(nullSubscription);
        assertThat(result, is(false));
    }
}

File: ClassyTaxiAppKotlin/app/src/test/java/com/example/subscriptions/billing/BillingUtilitiesTest.kt
package com.example.subscriptions.billing

import com.example.subscriptions.data.SubscriptionStatus
import org.hamcrest.CoreMatchers.`is`
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.nullValue
import org.junit.Test


class BillingUtilitiesTest {

    @Test
    fun subscriptionForSku_returnsSubscriptionForExistingSku() {
        val subscriptionList = listOf(
            SubscriptionStatus(sku = "basic_subscription"),
            SubscriptionStatus(sku = "premium_subscription")
        )
        val sku = "basic_subscription"
        val result = subscriptionForSku(subscriptionList, sku)
        assertThat(sku, `is`(equalTo(result?.sku)))
    }

    @Test
    fun subscriptionForSku_returnsNullForNonExistingSku() {
        val noSkuSubscriptionList = listOf(
            SubscriptionStatus(sku = "basic_subscription"),
            SubscriptionStatus(sku = "premium_subscription")
        )
        val sku = "non_cheap_subscription"
        val result = subscriptionForSku(noSkuSubscriptionList, sku)
        assertThat(result, `is`(nullValue()))
    }

    @Test
    fun isGracePeriod_returnsTrueIfInGracePeriod() {
        val onGraceSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            isGracePeriod = true
        )
        val result = isGracePeriod(onGraceSubscription)
        assertThat("Grace period is shown", result, `is`(equalTo(true)))
    }

    @Test
    fun isGracePeriod_returnsFalseIfNotInGracePeriod() {
        val notOnGraceSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            isGracePeriod = false
        )
        val result = isGracePeriod(notOnGraceSubscription)
        assertThat("Grace period is not shown", result, `is`(equalTo(false)))
    }

    @Test
    fun isGracePeriod_returnsFalseIfAlreadyOwnedSub() {
        val alreadyOwnedSubscription = SubscriptionStatus(
            subAlreadyOwned = true,
            isEntitlementActive = true,
            isGracePeriod = true
        )
        val result = isGracePeriod(alreadyOwnedSubscription)
        assertThat("Grace period is not shown for already owned subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isGracePeriod_returnsFalseIfEntitlementInactive() {
        val entitlementInactiveSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = false,
            isGracePeriod = true
        )
        val result = isGracePeriod(entitlementInactiveSubscription)
        assertThat("Grace period is not shown for inactive entitlement subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isGracePeriod_returnsFalseForNullSub() {
        val nullSubscription = null
        val result = isGracePeriod(nullSubscription)
        assertThat("Grace period is not shown due to null subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isSubscriptionRestore_returnsTrueForNonRenewingSub() {
        val willNotRenewSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            willRenew = false
        )
        val result = isSubscriptionRestore(willNotRenewSubscription)
        assertThat("Restore option is shown for subscription that will not be renewed", result, `is`(equalTo(true)))
    }

    @Test
    fun isSubscriptionRestore_returnsFalseForRenewingSub() {
        val willRenewSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            willRenew = true
        )
        val result = isSubscriptionRestore(willRenewSubscription)
        assertThat("Restore option is not shown for subscription that will be renewed", result, `is`(equalTo(false)))
    }

    @Test
    fun isSubscriptionRestore_returnsFalseForNullSub() {
        val nullSubscription = null
        val result = isSubscriptionRestore(nullSubscription)
        assertThat("Subscription restore is not shown due to null subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isBasicContent_returnsTrueForBasicSku() {
        val basicSkuSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            sku = "basic_subscription"
        )
        val result = isBasicContent(basicSkuSubscription)
        assertThat("Basic content is shown", result, `is`(equalTo(true)))
    }

    @Test
    fun isBasicContent_returnsFalseForNonBasicSku() {
        val nonBasicSkuSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            sku = "premium_subscription"
        )
        val result = isBasicContent(nonBasicSkuSubscription)
        assertThat("Basic content is not shown", result, `is`(equalTo(false)))
    }

    @Test
    fun isBasicContent_returnsFalseForNullSub() {
        val nullSubscription = null
        val result = isBasicContent(nullSubscription)
        assertThat("Basic content is not shown due to null subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isPremiumContent_returnsTrueForPremiumSku() {
        val premiumSkuSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            sku = "premium_subscription"
        )
        val result = isPremiumContent(premiumSkuSubscription)
        assertThat("Premium content is shown", result, `is`(equalTo(true)))
    }

    @Test
    fun isPremiumContent_returnsFalseForNonPremiumSku() {
        val nonPremiumSkuSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            sku = "basic_subscription"
        )
        val result = isPremiumContent(nonPremiumSkuSubscription)
        assertThat("Premium content is not shown", result, `is`(equalTo(false)))
    }

    @Test
    fun isPremiumContent_returnsFalseForNullSub() {
        val nullSubscription = null
        val result = isPremiumContent(nullSubscription)
        assertThat("Premium content is not shown due to null subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isAccountHold_returnsTrueForOnHoldSub() {
        val onHoldSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = false,
            isAccountHold = true
        )
        val result = isAccountHold(onHoldSubscription)
        assertThat("Account hold is shown", result, `is`(equalTo(true)))
    }

    @Test
    fun isAccountHold_returnsFalseForNonHoldSub() {
        val notHoldSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            isAccountHold = false
        )
        val result = isAccountHold(notHoldSubscription)
        assertThat("Account hold is not shown", result, `is`(equalTo(false)))
    }

    @Test
    fun isAccountHold_returnsFalseForNullSub() {
        val nullSubscription = null
        val result = isAccountHold(nullSubscription)
        assertThat("Account hold is not shown due to null subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isPaused_returnsTrueForPausedSub() {
        val pausedSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = false,
            isPaused = true
        )
        val result = isPaused(pausedSubscription)
        assertThat("Account paused is shown", result, `is`(equalTo(true)))
    }

    @Test
    fun isPaused_returnsFalseForNonPausedSub() {
        val pausedSubscription = SubscriptionStatus(
            subAlreadyOwned = false,
            isEntitlementActive = true,
            isPaused = false
        )
        val result = isPaused(pausedSubscription)
        assertThat("Account paused is not shown", result, `is`(equalTo(false)))
    }

    @Test
    fun isPaused_returnsFalseForNullSub() {
        val nullSubscription = null
        val result = isPaused(nullSubscription)
        assertThat("Account paused is not shown due to null subscription", result, `is`(equalTo(false)))
    }

    @Test
    fun isTransferRequired_returnsTrueForAlreadyOwnedSub() {
        val alreadyOwnedSubscription = SubscriptionStatus(subAlreadyOwned = true)
        val result = isTransferRequired(alreadyOwnedSubscription)
        assertThat("Transfer is required due to sub being already owned", result, `is`(equalTo(true)))
    }

    @Test
    fun isTransferRequired_returnsFalseForNotAlreadyOwnedSub() {
        val notAlreadyOwnedSubscription = SubscriptionStatus(subAlreadyOwned = false)
        val result = isTransferRequired(notAlreadyOwnedSubscription)
        assertThat("Transfer is not required", result, `is`(equalTo(false)))
    }

    @Test
    fun isTransferRequired_returnsFalseForNullSub() {
        val nullSubscription = null
        val result = isTransferRequired(nullSubscription)
        assertThat("Transfer is not required due null sub", result, `is`(equalTo(false)))
    }
}